### DISCUSSION TITLE
circular dependency

### DISCUSSION THREAD

**Discussion by Alessandro-Barbieri** [sagemathinc/cocalc#5521] [General]
Is this circular dependency needed?
https://github.com/sagemathinc/cocalc/blob/154d2152bada8f1231b34077df3d941472f2fcd7/src/packages/util/package.json#L32

**Comment by williamstein** [comment]
Yes.  It makes it so we can do  things like this (in packages/util/sync/editor/generic/util.ts):
```ts
import { cmp_array } from "@cocalc/util/misc";
```
anywhere in the codebase, in other modules and also directly in this module.  This is often better than doing relative importants like this, which are error prone and have be changed when a file is moved:
```ts
import {cmp_array} from "../../../misc";
```
I know it is also possible (https://www.typescriptlang.org/tsconfig#baseUrl) for typescript code to use the baseUrl option in tsconfig.json to accomplish something similar though less robust, across a monorepo.

**Comment by williamstein** [comment]
One other comment is that the CoCalc assumes you're using a recent version of npm, and definitely doesn't support yarn or npm <= 6.x

**Comment by williamstein** [comment]
I've eliminated these circular dependencies.   https://github.com/sagemathinc/cocalc/commit/5b2a6ee9230e544c379986364149bacdef89d694
